temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+                temporalio.activity._Context(
+                    info=lambda: info,
+                    heartbeat=None,
+                    cancelled_event=running_activity.cancelled_event,
+                    worker_shutdown_event=running_activity.worker_shutdown_event,
+                )
+            )
+            temporalio.activity.logger.debug("Starting activity")
+
+            # Build the interceptors chaining in reverse. We build a context right
+            # now even though the info() can't be intercepted and heartbeat() will
+            # fail. The interceptors may want to use the info() during init.
+            impl: ActivityInboundInterceptor = _ActivityInboundImpl()
+            for interceptor in reversed(list(self._config["interceptors"])):
+                impl = interceptor.intercept_activity(impl)
+            # Init
+            impl.init(_ActivityOutboundImpl(self, info))
+            # Exec
+            result = await impl.execute_activity(input)
+            # Convert result if not none. Since Python essentially only supports
+            # single result types (even if they are tuples), we will do the
+            # same.
+            if result is None:
+                completion.result.completed.SetInParent()
+            else:
+                result_payloads = await self._config["client"].data_converter.encode(
+                    [result]
+                )
+                # We have to convert from Temporal API payload to Core payload
+                completion.result.completed.result.metadata.update(
+                    result_payloads[0].metadata
+                )
+                completion.result.completed.result.data = result_payloads[0].data
+        except (BaseException, asyncio.CancelledError) as err:
+            try:
+                if isinstance(err, temporalio.activity.CompleteAsyncError):
+                    temporalio.activity.logger.debug("Completing asynchronously")
+                    completion.result.will_complete_async.SetInParent()
+                elif (
+                    isinstance(err, asyncio.CancelledError)
+                    and running_activity.cancelled_due_to_heartbeat_error
+                ):
+                    err = running_activity.cancelled_due_to_heartbeat_error
+                    temporalio.activity.logger.debug(
+                        f"Completing as failure during heartbeat with error of type {type(err)}: {err}",
+                        extra={"error": err},
+                        exc_info=True,
+                    )
+                    await temporalio.exceptions.apply_exception_to_failure(
+                        err,
+                        self._config["client"].data_converter,
+                        completion.result.failed.failure,
+                    )
+                elif (
+                    isinstance(err, asyncio.CancelledError)
+                    and running_activity.cancelled_by_request
```

Comment:
Note that core might cancel due to various reasons like workflow already complete

Reply:
I just meant that cancellation is also a mechanism for core to propagate heartbeat errors back to the activity.
The fact that core cancelled the activity doesn't mean it completes as cancelled and might need to be considered as failure (I think we overlooked that in TS too - we didn't use cancels like we do now when that code was written).

Reply:
I see that core sends a cancel with one of three enum reasons and the non-cancelled-request ones won't do anything with the error anyways (activity already considered handled elsewhere). So is the concern that I use the term `_by_request` here or that I should return a different error than cancelled error for a different reason? Can you clarify the suggestion?

Reply:
In node I ended up treating all of those as cancel requests.
I only treat cancel due to failed heartbeat converter as activity failure.

Reply:
Same here